index.js

```diff
@@ -192,6 +193,93 @@ module.exports = class HyperDHT extends DHT {
     }
   }
 
+  async immutableGet (key) {
+    if (Buffer.isBuffer(key) === false) throw new Error('key must be a buffer')
+    const query = this.query(key, 'immutable_get', null, { map: mapImmutable })
+    const check = Buffer.allocUnsafe(32)
+    for await (const node of query) {
+      const { value } = node
+      sodium.crypto_generichash(check, value)
+      if (check.equals(key)) return node
+    }
+    throw Error('not found')
+  }
+
+  async immutablePut (value) {
+    if (Buffer.isBuffer(value) === false) throw new Error('value must be a buffer')
+    if (value.length > PUT_VALUE_MAX_SIZE) {
+      throw new Error(`Value size must be <= ${PUT_VALUE_MAX_SIZE}`)
+    }
+    const key = Buffer.allocUnsafe(32)
+    sodium.crypto_generichash(key, value)
+    const query = this.query(key, 'immutable_get', null, {
+      map: mapImmutable,
+      commit (node, dht) {
+        return dht.request(key, 'immutable_put', value, node.from, { map: mapImmutable, token: node.token })
+      }
+    })
+    await query.finished()
+    return { key }
+  }
+
+  async mutableGet (key, { salt = null, seq = 0, latest = true } = {}) {
+    if (Buffer.isBuffer(key) === false) throw new Error('key must be a buffer')
+    if (typeof seq !== 'number') throw new Error('seq should be a number')
+    if (salt) {
+      if (Buffer.isBuffer(salt) === false) throw new Error('salt must be a buffer')
+      if (salt.length > 64) throw new Error('salt size must be no greater than 64 bytes')
+    }
+
+    const request = cenc.encode(messages.mutableReq, { salt, seq })
+
+    const query = this.query(key, 'mutable_get', request, { map: mapMutable })
+    const userSeq = seq
+    let topSeq = seq
+    let result = null
+    for await (const node of query) {
+      const { value: item, id, ...meta } = node
+      const { value, signature, seq: storedSeq, publicKey } = cenc.decode(messages.mutableRes, item)
+      const msg = hypersign.signable(value, { salt, seq: storedSeq })
+      if (storedSeq >= userSeq && sodium.crypto_sign_verify_detached(signature, msg, publicKey)) {
+        if (latest === false) return { id, value, signature, seq: storedSeq, salt, ...meta }
+        if (storedSeq >= topSeq) {
+          topSeq = storedSeq
+          result = { id, value, signature, seq: storedSeq, salt, ...meta }
+        }
+      }
+    }
+    return result
+  }
+
+  async mutablePut (value, opts = {}) {
+    if (Buffer.isBuffer(value) === false) throw new Error('value must be a buffer')
+    if (value.length > PUT_VALUE_MAX_SIZE) {
+      throw new Error(`Value size must be <= ${PUT_VALUE_MAX_SIZE}`)
+    }
+    const { seq = 0, salt = null, keyPair, signature = hypersign.sign(value, { ...opts, keypair: keyPair }) } = opts
```

Comment:
this one is bit out of my paygrade i think, what's going on?

Reply:
ahha, it's just auto generating a signature from the keypair if signature is not supplied 

I'll move the signing into dht and then we can revisit